File: java/account/AccountImpl.java
/**
 * @file AccountDelegate.java
 * @author Ronald T. Fernandez
 * @version 1.0
 */
package account;

import corbaAccount.Account;
import corbaAccount.AccountPOA;
import corbaAccount.Operation;
import corbaAccount.date;
import corbaAccount.operationType;

import connection.Connection;

/**
 * This class is an implementation of the Account CORBA interface
 */
class AccountImpl extends AccountPOA {

	/**
	 * Account identifier
	 */
	private int _accountId;
	/**
	 * Account owner name
	 */
	private String _name;
	/**
	 * Account owner surname
	 */
	private String _surname;
	/**
	 * Account creation date
	 */
	date _dateAccountCreated;
	/**
	 * Updated account balance
	 */
	private float _balance;
	/**
	 * Operations performed in the account
	 */
	private Operation[] _accountOperations = null;
	/**
	 * CORBA connection, used to publish the current instance in the naming service (if proceed)
	 */
	private Connection _connection;
	
	/** 
	 * Class constructor
	 * @param a		Account to be copied
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountImpl(Account a) throws Exception {
		super();
		accountId(a.accountId());
		name(a.name());
		surname(a.surname());
		dateAccountCreated(a.dateAccountCreated());
		balance(a.balance());
		accountOperations(a.accountOperations());
		
		// Activate the CORBA object
		_connection = Connection.getInstance();
		_connection.activateServant(this);
	}
	
	/** 
	 * Class constructor
	 * @param name 				Account owner name
	 * @param surname			Account owner surname
	 * @param dateCreated		Account creation date
	 * @param balance			Initial account balance
	 * @param accOperations		Operations performed in the account
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountImpl(String name, String surname, date dateCreated, float balance, Operation[] accOperations) throws Exception {
		name(name);
		surname(surname);
		dateAccountCreated(dateCreated);
		balance(balance);
		accountOperations(accOperations);
		
		// Activate the CORBA object
		_connection = Connection.getInstance();
		_connection.activateServant(this);
	}
	
	/** 
	 * Class constructor
	 * @param name 				Account owner name
	 * @param surname			Account owner surname
	 * @param balance			Initial account balance
	 * @param accountId			Account identifier
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountImpl(String name, String surname, float balance, int accountId) throws Exception {
		super();
		accountId(accountId);
		name(name);
		surname(surname);
		balance(balance);
		accountOperations(new Operation[0]);
		
		// Activate the CORBA object
		_connection = Connection.getInstance();
		_connection.activateServant(this);
		
		DateDelegate dd = new DateDelegate();
		_dateAccountCreated = dd.getCorbaInstance();
	}
		
	/** 
	 * Class constructor
	 * @param name 				Account owner name
	 * @param surname			Account owner surname
	 * @param accountId			Account identifier
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountImpl(String name, String surname, int accountId) throws Exception {
		this(name, surname, 0.0f, accountId);
	}
	
	/**
	 * Get the accountId
	 * @return accountId
	 */
	@Override
	public int accountId() {
		return _accountId;
	}

	/**
	 * Set the accountId
	 * @param	newAccountId	New accountId
	 */
	@Override
	public void accountId(int newAccountId) {
		_accountId = newAccountId;
	}

	/**
	 * Get the account owner name
	 * @return Owner name
	 */
	@Override
	public String name() {
		return _name;
	}

	/**
	 * Set the account owner name
	 * @param	newName	New owner name
	 */
	@Override
	public void name(String newName) {
		_name = newName;	
	}

	/**
	 * Get the account owner surnamed
	 * @return Owner surname
	 */
	@Override
	public String surname() {
		return _surname;
	}

	/**
	 * Set the account owner surname
	 * @param	newSurname	New owner surname
	 */
	@Override
	public void surname(String newSurname) {
		_surname = newSurname;	
	}

	/**
	 * Get the account creation date
	 * @return Account creation date
	 */
	@Override
	public date dateAccountCreated() {
		return _dateAccountCreated;
	}

	/**
	 * Set the account creation date
	 * @param	newDateAccountCreated	Updated creation date
	 */
	@Override
	public void dateAccountCreated(date newDateAccountCreated) {
		_dateAccountCreated = newDateAccountCreated;
	}

	/**
	 * Get the account balance
	 * @return balance
	 */
	@Override
	public float balance() {
		return _balance;
	}

	/**
	 * Set the account balance
	 * @param	newBalance	Account balance
	 */
	@Override
	public void balance(float newBalance) {
		_balance = newBalance;
	}

	/**
	 * Get the list of account operations in the account
	 * @return Array of accountOperations
	 */
	@Override
	public Operation[] accountOperations() {
		return _accountOperations;
	}

	/**
	 * Set the list of operations performed in the account.
	 * NNOTE: This overrides the existing accounts in case they exist!
	 * @param	newAccountOperations	List of account operations
	 */
	@Override
	public void accountOperations(Operation[] newAccountOperations) {
		_accountOperations = newAccountOperations;
		// Update the account balance
		for (Operation op : newAccountOperations) {
			if (op.type() == operationType.WITHDRAW)
				_balance -= op.amount();
			else
				_balance += op.amount();
		}
	}

	/**
	 * Get the account owner full name
	 * @return Concatenation os name and surname
	 */
	@Override
	public String details() {
		return _name + " " + _surname;
	}

	/**
	 * Add an operation and update the account balance
	 * @param	op	Operation to be added
	 */
	@Override
	public void addOperation(Operation op) {
		Operation[] tmp = new Operation[_accountOperations.length + 1];
		int pos = 0;
		// Update the operations performed in the account
		for (Operation o : _accountOperations) {
			tmp[pos++] = o;
		}
		tmp[pos] = op;
		_accountOperations = tmp;
		
		// Update balance
		if (op.type() == operationType.WITHDRAW)
			_balance -= op.amount();
		else
			_balance += op.amount();
	}

	/**
	 * Get a summary of the account
	 * @return String with the details of the account
	 */
	@Override
	public String _toString() {
		return "accountId: " + String.format("%2s",Integer.valueOf(_accountId)).replaceAll(" " , "0") + 
				", owner: " + _name + " " + _surname + ", balance: " + _balance + ", operations: " + _accountOperations.length;
	}
	
	/**
	 * Get the CORBA instance associated to this account
	 * @return Account CORBA instance
	 */
	public Account getCorbaInstance() {
		return _this();
	}
	
}

File: cxx/account/AccountImpl.cpp
/**
 * @file AccountImpl.cpp
 * @author Ronald T. Fernandez
 * @version 1.0
 */

#include "AccountImpl.h"

#include <exception>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>

namespace account {

std::shared_ptr<connection::Connection> AccountImpl::_connection = nullptr;

AccountImpl::~AccountImpl() {
	_connection->deactivateServant(this);
}

AccountImpl::AccountImpl(corbaAccount::Account_ptr a) : _accountId(a->accountId()), _name(a->name()), _surname(a->surname()), _balance(a->balance()) {
	dateAccountCreated(a->dateAccountCreated());
	accountOperations(*a->accountOperations());

	// Activate the CORBA object
	_connection = connection::Connection::getInstance();
	_connection->activateServant(this);
}

AccountImpl::AccountImpl(std::string name, std::string surname, corbaAccount::date_ptr dateAccCreated, float balance, corbaAccount::accountOperationsType& accOperations, unsigned int accountId) :
	_accountId(accountId), _name(name), _surname(surname), _balance(balance){
	dateAccountCreated(dateAccCreated);
	accountOperations(accOperations);

	// Activate the CORBA object
	_connection = connection::Connection::getInstance();
	_connection->activateServant(this);
}


AccountImpl::AccountImpl(std::string name, std::string surname, float balance, unsigned int accountId) : _accountId(accountId), _name(name), _surname(surname),
		_balance(balance), _accountOperations(){

	// Activate the CORBA object
	_connection = connection::Connection::getInstance();
	_connection->activateServant(this);

	// Initialize the account creation date
	_dateAccountCreated = new DateDelegate();
}

AccountImpl::AccountImpl(std::string name, std::string surname, unsigned int accountId) : _accountId(accountId), _name(name), _surname(surname), _balance(0.0f) {
	// Initializes the connection (for future references)
	_connection = connection::Connection::getInstance();
}

::CORBA::Long AccountImpl::accountId() {
	return _accountId;
}

void AccountImpl::accountId(::CORBA::Long _v) {
	_accountId = _v;
}
char* AccountImpl::name() {
	return CORBA::string_dup(_name.c_str());
}

void AccountImpl::name(const char* _v) {
	_name = CORBA::string_dup(_v);
}

char* AccountImpl::surname() {
	return CORBA::string_dup(_surname.c_str());
}

void AccountImpl::surname(const char* _v) {
	_surname = CORBA::string_dup(_v);
}

corbaAccount::date_ptr AccountImpl::dateAccountCreated() {
	return _dateAccountCreated->getCorbaInstance();
}

void AccountImpl::dateAccountCreated(::corbaAccount::date_ptr _v) {
	if (CORBA::is_nil(_v))
		_dateAccountCreated = new DateDelegate(_v->year(), _v->month(), _v->day());
	else {
		std::cerr << "Cannot set a NULL date to the account" << std::endl;
		throw std::exception();
	}
}

::CORBA::Float AccountImpl::balance() {
	return _balance;
}

void AccountImpl::balance(::CORBA::Float _v) {
	_balance = _v;
}

corbaAccount::accountOperationsType* AccountImpl::accountOperations() {
	return &_accountOperations;
};

void AccountImpl::accountOperations(const ::corbaAccount::accountOperationsType& _v) {
	_accountOperations = _v;
	// Update the account balance
	for (size_t pos=0; pos < _accountOperations.length(); ++pos) {
		if (_accountOperations[pos]->type() == corbaAccount::operationType::WITHDRAW)
			_balance -= _accountOperations[pos]->amount();
		else
			_balance -= _accountOperations[pos]->amount();
	}
}

char* AccountImpl::details() {
	return CORBA::string_dup((_name + " " + _surname).c_str());
};

void AccountImpl::addOperation(::corbaAccount::Operation_ptr op) {

	if (CORBA::is_nil(op)) {
		std::cerr << "Cannot add a NULL operation" << std::endl;
		throw std::exception();
	}

	size_t pos = _accountOperations.length();
	_accountOperations.length(pos + 1);
	_accountOperations[pos] = op;
	// Update the account balance
	if (op->type() == corbaAccount::operationType::WITHDRAW)
		_balance -= op->amount();
	else
		_balance += op->amount();
};

char* AccountImpl::toString() {
	std::ostringstream os;
	os << std::setw(2) << std::setfill('0') << "accountId: " << _accountId << ", owner: " << _name << " " << _surname << ", balance: " <<
			_balance << ", operations: " << _accountOperations.length();
	return CORBA::string_dup(os.str().c_str());
};
};

